Add support for generating Jupyterhub links to alias files

JamesWrigley opened the pull request:
This required tracking any alias files used through the `DataCollection` object.

The result is:
![image](https://github.com/user-attachments/assets/75c2ce99-4724-4768-9fd5-ec8561f7ada5)

takluyver commented on extra_data/reader.py:
```diff
@@ -629,6 +631,10 @@ def _merge_aliases(self, alias_dicts):
 
+    def _merge_alias_files(self, *alias_files):
+        all_files = chain.from_iterable(alias_files)
+        return list(set(all_files))
```
```suggestion
        return sorted(set(all_files))
```

Sets give a random order, which doesn't really matter, but it's always nice to be deterministic if possible.

JamesWrigley replied:
Fixed in d92ccd1.

takluyver commented on extra_data/reader.py:
```diff
@@ -693,10 +702,11 @@ def is_valid_alias(k, v):
                 alias_dicts.append(alias_def)
             elif isinstance(alias_def, (str, os.PathLike)):
                 # From a file.
+                alias_files.append(alias_def)
```
Do we want to normalise the type here?

```suggestion
                alias_files.append(os.fspath(alias_def))
```

This would convert Path objects to strings.

philsmt replied:
I was about to comment, I would prefer it we always have `str` or `Path` internally (as well as externally via the property), and not return combination of types was passed.

takluyver commented on setup.py:
```diff
@@ -84,6 +84,7 @@ def find_version(*parts):
               'nbval',
               'pytest',
               'pytest-cov',
+              'pyyaml',
```
This is already in the regular requirements, so it doesn't need to be here.

```suggestion
```

JamesWrigley replied:
Whoops I missed that, fixed in d92ccd1.

takluyver reviewed (commented):
This broadly looks good, I have a few specific suggestions.

philsmt commented on extra_data/aliases.py:
```diff
@@ -167,6 +167,15 @@ def _resolve_aliased_selection(self, selection):
+
+    def jhub_links(self):
+        """Return clickable Jupyterhub links to any alias files that were used."""
+        paths_from_home = [str(p).replace("/gpfs", "GPFS") for p in self.file_paths()]
```
Actually, shouldn't we specifically exclude (and/or print a warning) for anything not under `/gpfs`? Someone else will hardly be able to open a jhub link pointing to someone's home directory beyond `~/GPFS`.

takluyver replied:
I'd assume the links are primarily for the person running the code, and if they can use an alias file in their home directory, they can follow the link for it as well. It's not ideal if they share the notebook with someone else, but using things in your home directory harms that anyway.

philsmt commented on extra_data/reader.py:
```diff
@@ -117,6 +118,7 @@ def __init__(
             }
         self._sources_data = sources_data
 
+        self._alias_files = [] if alias_files is None else alias_files
```
It makes sense in the context of this PR, but could you add a comment this is only for tracking purposes? The actual aliases are still carried via the `aliases` dict after all.

philsmt commented:
Generally good, also got some comments about details.

Interestingly it's a situation I have not yet encountered, but I trust it's relevant enough for you.

JamesWrigley commented on extra_data/tests/conftest.py:
```diff
@@ -5,11 +5,19 @@
+# element that makes it impossible to deserialize without a custom constructor.
+def tuple_representer(dumper, data):
+    return dumper.represent_list(list(data))
+yaml.add_representer(tuple, tuple_representer)
```
Yep that works, fixed in d92ccd1.

JamesWrigley commented:
Yeah I added it because people frequently want to add more aliases and have trouble finding the file they're defined in. To that end, I made another change in d92ccd1 to show the links in `repr()` as well to be even more obvious:
<img width="1017" height="256" alt="image" src="https://github.com/user-attachments/assets/1d0bb1bd-842c-46d2-b2e8-34ffbe869d57" />

I also changed the link to use Jupyterlab instead of Ye Olde Notebooks since that's what most people seem to use these days.

takluyver commented:
Thanks, LGTM!
